File: Java/第3章 链表/3.3 带头结点的链表/SinglyLinkedList.java
public class SinglyLinkedList {
	private Node head;		// 链表头结点
	
	/**
	 * @brief  初始化链表
	 */
	public SinglyLinkedList() {
		this.head = null;
	}
	
	/**
	 * @brief  判断是否是空链表
	 * @retval 空链表返回true，为空链表返回false
	 */
	public boolean isEmpty() {
	    return this.head == null;
	}

	/**
	 * @brief  获取链表长度
	 * @retval 链表长度
	 */
	public int getLength() {
	    Node temp = this.head;
	    int len = 0;
	    while(temp != null) {
	    	len++;
	    	temp = temp.next;
	    }
	    return len;
	}
	
	/**
	 * @brief  打印链表
	 */
	public void show() {
	    Node temp = this.head;
	    while(temp != null) {
	        System.out.print(temp.data + " ");
	    	temp = temp.next;
	    }
	    System.out.println();
	}
	
	/**
	 * 头部插入结点
	 * @param val
	 */
	public void addFront(int val) {
		Node newNode = new Node(val);
		newNode.next = this.head;
		this.head = newNode;
	}
	
	/**
	 * @brief  插入结点
	 * @param  pos: 插入位置
	 * @param  val: 插入元素
	 */
	public void insert(int pos, int val) {
	    // 超出链表范围
	    if(pos < 0 || pos > getLength()) {
	        System.err.println("Error: position out of bound.");
	    	return;
		}
	    
	    // 头部插入
	    if(pos == 0) {
	    	addFront(val);
	    	return;
	    }
	    
	    // 找到要插入结点的前一个结点
	    Node temp = this.head;
	    for(int i = 0; i < pos - 1; i++) {
	    	temp = temp.next;
	    }
	    Node newNode = new Node(val);
	    newNode.next = temp.next;
	    temp.next = newNode;
	}
	
	/**
	 * @brief  删除结点
	 * @param  pos: 插入位置
	 */
	public void delete(int pos) {
	    // 超出链表范围
	    if(pos < 0 || pos >= getLength()) {
	    	System.err.println("Error: position out of bound.");
			return;
		}
	    
	    // 删除头结点
	    if(pos == 0) {
	    	this.head = this.head.next;
	    	return;
	    }
	    
	    // 找到删除位置
	    Node temp = this.head;
	    for(int i = 0; i < pos - 1; i++) {
	    	temp = temp.next;
	    }
	    temp.next = temp.next.next;
	}

	/**
	 * @brief  查找元素
	 * @param  val: 查找值
	 * @retval 返回结点指针，未找到返回null
	 */
	public Node search(int val) {
	    // 查找元素位置
	    Node temp = this.head;
	    while(temp != null) {
	    	if(temp.data == val) {
	    		return temp;
	    	}
	        temp = temp.next;
	    }
	    return null;
	}

	/**
	 * @brief  修改元素
	 * @param  pos: 修改位置
	 * @param  val - 新数据
	 */
	public void set(int pos, int val) {
	    // 超出链表范围
	    if(pos < 0 || pos >= getLength()) {
	    	System.err.println("Error: position out of bound.");
			return;
		}

	    // 找到元素位置
		Node temp = this.head;
	    for(int i = 0; i < pos; i++) {
	        temp = temp.next;
	    }
	    temp.data = val;
	}
	
	public static void main(String[] args) {
		SinglyLinkedList list = new SinglyLinkedList();
		
		list.insert(0, 1);			// [1]
		list.insert(0, 2);			// [2, 1]
		list.insert(2, 3);			// [2, 1, 3]
		list.show();
		
		list.delete(0);				// [1, 3]
		list.insert(1, 4);			// [1, 4, 3]
		list.delete(2);				// [1, 4]
		list.show();
		
		list.set(0, 5);				// [5, 4]
		list.set(1, 6);				// [5, 6]
		list.show();
		
		System.out.println(list.search(6).data);	// 6
	}
}

File: code/Python/Chapter3/3-1/singly_linked_list.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.__head = None
        self.__tail = None
        self.__len = 0
    
    def is_empty(self):
        return self.__head == None
    
    def length(self):
        return self.__len
    
    def add(self, data):
        node = Node(data)
        if self.is_empty():
            self.__head = node
        else:
            self.__tail.next = node
        self.__tail = node
        self.__len += 1
    
    def insert(self, index, data):
        if index < 0 or index > self.__len:
            raise IndexError("Index out of range")

        node = Node(data)
        if index == 0:
            node.next = self.__head
            self.__head = node
        else:
            cur = self.__head
            for _ in range(index - 1):
                cur = cur.next
            node.next = cur.next
            cur.next = node
        self.__len += 1
    
    def remove(self, index):
        if index < 0 or index >= self.__len:
            raise IndexError("Index out of range")

        if index == 0:
            data = self.__head.data
            self.__head = self.__head.next
        else:
            cur = self.__head
            for _ in range(index - 1):
                cur = cur.next
            data = cur.next.data
            cur.next = cur.next.next
        self.__len -= 1
        return data
    
    def get(self, index):
        if index < 0 or index >= self.__len:
            raise IndexError("Index out of range")

        cur = self.__head
        for _ in range(index):
            cur = cur.next
        return cur.data
    
    def __str__(self):
        cur = self.__head
        s = ""
        while cur != None:
            s += str(cur.data) + " "
            cur = cur.next
        return s
    
    def __iter__(self):
        self.__cur = self.__head
        return self
    
    def __next__(self):
        if self.__cur == None:
            raise StopIteration
        data = self.__cur.data
        self.__cur = self.__cur.next
        return data